src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -35311,6 +35330,92 @@ static void Main()
+}
+[NonNullTypes(true)]
+class B4 : A<string?>
+{
+    public override void F1<U>() { }
+    public override void F2<U>() { }
+    public override void F3<U>() { }
+    public override void F4<U>() { }
+}";
+            var comp = CreateCompilation(source, references: new[] { new CSharpCompilationReference(comp0) }, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics();
+            verifyAllConstraintTypes();
+
+            comp = CreateCompilation(source, references: new[] { comp0.EmitToImageReference() }, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics();
+            verifyAllConstraintTypes();
+
+            void verifyAllConstraintTypes()
+            {
+                verifyConstraintTypes("B1.F1", "System.String", "I<System.String>");
+                verifyConstraintTypes("B1.F2", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B1.F3", "System.String", "I<System.String>");
+                verifyConstraintTypes("B1.F4", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B2.F1", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B2.F2", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B2.F3", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B2.F4", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B3.F1", "System.String!", "I<System.String!>");
+                verifyConstraintTypes("B3.F2", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B3.F3", "System.String!", "I<System.String!>"); // PROTOTYPE(NullableReferenceTypes): Should be I<System.String!>!. See TypeSymbolWithAnnotations.SubstituteType.
+                verifyConstraintTypes("B3.F4", "System.String?", "I<System.String?>"); // PROTOTYPE(NullableReferenceTypes): Should be I<System.String?>!. See TypeSymbolWithAnnotations.SubstituteType.
+                verifyConstraintTypes("B4.F1", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B4.F2", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B4.F3", "System.String?", "I<System.String?>"); // PROTOTYPE(NullableReferenceTypes): Should be I<System.String?>!. See TypeSymbolWithAnnotations.SubstituteType.
+                verifyConstraintTypes("B4.F4", "System.String?", "I<System.String?>"); // PROTOTYPE(NullableReferenceTypes): Should be I<System.String?>!. See TypeSymbolWithAnnotations.SubstituteType.
+            }
+
+            void verifyConstraintTypes(string methodName, params string[] expectedTypes)
+            {
+                var constraintTypes = comp.GetMember<MethodSymbol>(methodName).TypeParameters[0].ConstraintTypesNoUseSiteDiagnostics;
+                AssertEx.Equal(expectedTypes, constraintTypes.SelectAsArray(t => t.ToTestDisplayString(true)));
+            }
+        }
```

Comment:
>} [](start = 8, length = 1)

Could you also add a test for the constraint type unification error? `where T : I<string?>, I<T>` and some variants. 
We should probably ignore nullability there too, and produce diagnostics as today in such scenarios.

Reply:
There are `TypeUnification_0*` tests. Are there additional variants that should be covered?